### ISSUE TITLE
rfnoc_image_builder crashes in python 3.5.2

### ISSUE THREAD

**Issue by jelder19** [EttusResearch/uhd#417]
## Issue Description ##
When running rfnoc_image_builder using python 3.5.2, crashes with the following error:
*** TypeError: '_sre.SRE_Match' object is not subscriptable
line 260 of image_builder.py -> if ident[0] in block["parameters"]:

## Setup Details ##
UHD latest commit of master branch
Ubuntu 16.04
python 3.5.2

## Expected Behavior ##
works

## Actual Behaviour ##
crashes

## Steps to reproduce the problem ##
cd ~/uhd/fpga/usrp3/top/x300
rfnoc_image_builder -y ./x310_rfnoc_image_core.yml -t X310_XG

## Additional Information ##
Turns out the regex module behavior changed in 3.7. in 3.5, you can't subscript the match result.
Replacing 'ident[0]' with 'ident.group(0)' on lines 260-262 fixes the problem for me.
I'm not sure if this works in 3.7+, so you guys will have to figure that out.

**Comment by mbr0wn** [comment]
@jelder19 many thanks for the bug report. Using `group(0)` is correct and works fine across version. It is the right answer. We'll apply this fix soon.

**Comment by mbr0wn** [comment]
@jelder19 we actually found a bunch more issues related to Python 3.5. However, we also decided to no longer support 3.5 in the upcoming release (UHD 4.1), among other things because we really want to start using some Python 3.6 features.

Here are the patches that will let you run UHD with Python 3.5:
```patch
From 570c377db41e1350167a6ede76ea1784ecc595ad Mon Sep 17 00:00:00 2001
From: Martin Braun <martin.braun@ettus.com>
Date: Mon, 28 Jun 2021 13:12:27 +0200
Subject: [PATCH 1/2] uhd: Remove Python 3.6-isms

As much as it pains my dear heart, we currently promise Python 3.5
support in our dependency list. This removes some usage of f-strings
from UHD which currently break Py3.5-support.
---
 host/lib/include/uhdlib/usrp/common/rpc.py    |  4 ++--
 host/python/uhd/usrp/cal/switch.py            |  5 +++--
 host/python/uhd/usrp/cal/usrp_calibrator.py   |  6 +++--
 host/tests/devtest/multi_usrp_test.py         | 22 +++++++++++--------
 .../run_X4xx_max_rate_tests.py                |  4 ++--
 host/tests/verify_fbs_test.py                 |  2 +-
 6 files changed, 25 insertions(+), 18 deletions(-)

diff --git a/host/lib/include/uhdlib/usrp/common/rpc.py b/host/lib/include/uhdlib/usrp/common/rpc.py
index 4ca30b07d..88f266951 100644
--- a/host/lib/include/uhdlib/usrp/common/rpc.py
+++ b/host/lib/include/uhdlib/usrp/common/rpc.py
@@ -14,13 +14,13 @@ class Function:
         self.does_return = return_type != "void"
         self.return_type = return_type
         self.arg_names = [arg[1] for arg in args]
-        self.rpcname = f"\"{function_name}\""
+        self.rpcname = "\"{}\"".format(function_name)
         self.args = [" ".join(arg) for arg in args]
         self.has_rpcprefix = False
         self.no_claim = no_claim
 
     def enable_rpcprefix(self):
-        self.rpcname = f"_rpc_prefix + \"{self.name}\""
+        self.rpcname = "_rpc_prefix + \"{}\"".format(self.name)
         self.has_rpcprefix = True
 
 class Interface:
diff --git a/host/python/uhd/usrp/cal/switch.py b/host/python/uhd/usrp/cal/switch.py
index b17318b5e..4c255e4b5 100644
--- a/host/python/uhd/usrp/cal/switch.py
+++ b/host/python/uhd/usrp/cal/switch.py
@@ -53,8 +53,9 @@ class ManualSwitch(SwitchBase):
         if self.mode == 'auto':
             return # no need to wait for manual connection
         dev_type = "signal generator" if self.direction == 'rx' else "power meter"
-        input(f"[{self.direction}] Connect your {dev_type} to device channel {chan}, "
-              f"antenna {antenna}. Then, hit Enter.")
+        input("[{}] Connect your {} to device channel {}, "
+              "antenna {}. Then, hit Enter."
+              .format(self.direction, dev_type, chan, antenna))
 
 class NISwitch(SwitchBase):
     """
diff --git a/host/python/uhd/usrp/cal/usrp_calibrator.py b/host/python/uhd/usrp/cal/usrp_calibrator.py
index c129fb17d..2a20aaf85 100644
--- a/host/python/uhd/usrp/cal/usrp_calibrator.py
+++ b/host/python/uhd/usrp/cal/usrp_calibrator.py
@@ -307,7 +307,8 @@ class USRPCalibratorBase:
             # result of the equation is in dBm again. We omit the subtract-by-zero
             # since our variables don't have units.
             results[gain] = usrp_input_power - recvd_signal_power
-            self.log(f"{gain:4.2f} dB => {results[gain]:+6.2f} dBm")
+            self.log("{:4.2f} dB => {:+6.2f} dBm"
+                     .format(gain, results[gain]))
             # If we get too close to the noise floor, we stop
             if recvd_power - self._noise[freq][gain] <= 1.5:
                 self.log("Can no longer detect input signal. Terminating.")
@@ -323,7 +324,8 @@ class USRPCalibratorBase:
             self._usrp.set_tx_gain(gain, self._chan)
             time.sleep(0.1) # Settling time of the USRP, highly conservative
             results[gain] = self._meas_dev.get_power()
-            self.log(f"{gain:4.2f} dB => {results[gain]:+6.2f} dBm")
+            self.log("{:4.2f} dB => {:+6.2f} dBm"
+                     .format(gain, results[gain]))
         self.results[freq] = results
 
     def store(self):
diff --git a/host/tests/devtest/multi_usrp_test.py b/host/tests/devtest/multi_usrp_test.py
index 943733ba9..4eea373a1 100755
--- a/host/tests/devtest/multi_usrp_test.py
+++ b/host/tests/devtest/multi_usrp_test.py
@@ -38,11 +38,11 @@ class MethodExecutor(object):
         """
         self.tested.extend(method_names)
         method_names_str = ", ".join([method + "()" for method in method_names])
-        print(f"Executing methods: {method_names_str}")
+        print("Executing methods:", method_names_str)
         try:
             method_callback()
         except Exception as ex:
-            print(f"Error while executing methods: \n`{method_names_str}`:\n{ex}",
+            print("Error while executing methods: \n`{}`:\n{}".format(method_names_str, ex),
                   file=sys.stderr)
             self.failed.extend(method_names)
             return False
@@ -230,7 +230,8 @@ def list_test(usrp, prop, error_handling, post_hook=None, safe_values=None):
     # Try to set every possible value.
     for name in names:
         if safe_values and name not in safe_values:
-            print(f"Skipping value `{name}' for prop {prop}, considered unsafe.")
+            print("Skipping value `{}' for prop {}, considered unsafe."
+                  .format(name, prop))
             continue
         initial_value = getattr(usrp, getter)(0)
         try:
@@ -456,22 +457,25 @@ def power_test(usrp, direction, num_chans):
             except:
                 pass
             else:
-                raise(f"get_{direction}_power_range({chan}): "
-                      "Expected exception (no power cal), none occurred.")
+                raise("get_{}_power_range({}): "
+                      "Expected exception (no power cal), none occurred."
+                      .format(direction, chan))
             try:
                 get_cb(chan)
             except:
                 pass
             else:
-                raise(f"get_{direction}_power_reference({chan}): "
-                      "Expected exception (no power cal), none occurred.")
+                raise("get_{}_power_reference({}): "
+                      "Expected exception (no power cal), none occurred."
+                      .format(direction, chan))
             try:
                 set_cb(100, chan)
             except:
                 pass
             else:
-                raise(f"set_{direction}_power_reference({chan}): "
-                      "Expected exception (no power cal), none occurred.")
+                raise("set_{}_power_reference({}): "
+                      "Expected exception (no power cal), none occurred."
+                      .format(direction, chan))
             continue
         # pylint: enable=bare-except
         # Now check power API for reals:
diff --git a/host/tests/streaming_performance/run_X4xx_max_rate_tests.py b/host/tests/streaming_performance/run_X4xx_max_rate_tests.py
index ea424f14f..302bc7354 100644
--- a/host/tests/streaming_performance/run_X4xx_max_rate_tests.py
+++ b/host/tests/streaming_performance/run_X4xx_max_rate_tests.py
@@ -157,10 +157,10 @@ def main():
     args = parse_args()
 
     base_params = {
-        "args" : f"addr={args.addr},second_addr={args.second_addr}",
+        "args" : "addr={},second_addr={}".format(args.addr, args.second_addr),
     }
     if args.use_dpdk:
-        base_params["args"] += f",mgmt_addr={args.mgmt_addr},use_dpdk=1"
+        base_params["args"] += ",mgmt_addr={},use_dpdk=1".format(args.mgmt_addr)
 
     start_time = time.time()
 
diff --git a/host/tests/verify_fbs_test.py b/host/tests/verify_fbs_test.py
index 261037c4f..36d5f3db9 100644
--- a/host/tests/verify_fbs_test.py
+++ b/host/tests/verify_fbs_test.py
@@ -18,7 +18,7 @@ class VerifyFBSTest(unittest.TestCase):
     def test_flatbuffer_files(self):
         """ Check FBS file """
         utils_path = os.path.normpath(os.path.join(
-            pathlib.Path(__file__).parent.absolute(),
+            str(pathlib.Path(__file__).parent.absolute()),
             '..', 'utils', 'update_fbs.py'))
         print(utils_path)
         spec = importlib.util.spec_from_file_location("update_fbs", utils_path)
-- 
2.25.1
```

```patch
From 9d3c0fd7476fce30f6fe11a16a657bac43a1e5c3 Mon Sep 17 00:00:00 2001
From: Martin Braun <martin.braun@ettus.com>
Date: Thu, 17 Jun 2021 13:55:26 +0200
Subject: [PATCH 2/2] image builder: Fix Python 3.5 usage

This replaces a re matchobj index from [0] to group(0) for backward
compatibility with Python 3.5.
---
 host/python/uhd/imgbuilder/image_builder.py | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/host/python/uhd/imgbuilder/image_builder.py b/host/python/uhd/imgbuilder/image_builder.py
index 9ce57d0e0..a99a51076 100755
--- a/host/python/uhd/imgbuilder/image_builder.py
+++ b/host/python/uhd/imgbuilder/image_builder.py
@@ -259,9 +259,10 @@ class ImageBuilderConfig:
                             # place. That may result in an exception being
                             # thrown from eval(), but we'll catch it and
                             # report an error a bit later on.
-                            if ident[0] in block["parameters"]:
-                                val = str(block["parameters"][ident[0]])
-                                num_ports = re.sub(ident[0], val, num_ports)
+                            parameter = ident.group(0)
+                            if parameter in block["parameters"]:
+                                val = str(block["parameters"][parameter])
+                                num_ports = re.sub(parameter, val, num_ports)
 
                         # Now, with identifiers resolved to parameter values,
                         # attempt to evaluate the expression. If eval() fails,
-- 
2.25.1
```

**Comment by mbr0wn** [comment]
Closing, because we will be using Python 3.6 going forward.
